README.md
# easydata
Easy Data Program's source code

You need below packages for building program from source code. (use pip for installing these packages)
- beautifulsoup4
- XlsxWriter

You can download source files as zip file. (last version: V 0.5 Alpha)

Go to  the relese section for downloading binary release for Microsoft Windows


sourcecode/changelog.txt
EasyData From Chaharmahalmet By MiR.am
Program Developer: MiR.am (mir-am@hotmail.com)

-----------------------------------------------------------------------

Change Log:

Version 0.1 Alpha (June 8, 2015):

New Features:
- Downloading data from web page
- Parsing a HTML page (Thanks to BeautifulSoup4 module)
- Saving data from web page in text file
- Reading data from text file
- Extracting data from text file. Getting info such as station name, lat, lon, etc
- Process data reqeust by user
- Creating a Console Program (include 4 steps for getting data from chaharmahalmet)
- 8 Variables has been supported.
- 202 Stations has been supported.

Version 0.2 Alpha (June 11, 2015):

NewFeatures:
- Getting long-term data from chaharmahalmet and convert it to the excel file(Thanks to xlsx writer module.)
- Adding colored terminal text (Thanks to colorma module.)

SolvedIssues:
- Fixed invalid folder address by user.  


Version 0.3 Alpha (June 12, 2015):

New Features:
- Extracting Anually data from text files
- Converting Anually data from text files to excel file
- Adding new step so that user can choose options (Long-term data, Anually data, Monthly data)
- Adding functions for making dircetories and saving text files
- New function for processing several data request from user

Version 0.4 Alpha (June 14, 2015):

New Features:
- Extracting Monthly data from text files
- Converting Monthly data from text files to excel file
- Adding new Full Data options to user interface
- Extracting Full Data(all years + all months) from text files
- Converting Full Data from text files to excel files

SolvedIssues:
- Clearing Variables after each user request

Version 0.5 Alpha (July 11, 2015):

New Features:
- Graphical User Interface created for program. (Tkinter)

sourcecode/gui.py
__author__ = 'MiR.am'

# program name: EasyData
# developer: MiR.am
# date: July 11, 2015
# version: 0.5 Alpha
# License: GNU General Public License v3.0

# in this module, Graphical user interface created

# Tkinter module used for creating GUI
from Tkinter import *
from threading import Thread
import tkFileDialog
import tkMessageBox as tkmsg
import tkFont
import ctypes
import sys
import misc
import getdata


ctypes.windll.kernel32.SetConsoleTitleA("EasyData By MiR.am")

# information about program
about = """ EasyData Program\n
 Version: 0.5 Alpha (July 11, 2015)
 Program Developer: MiR.am (mir-am@hotmail.com)
 License: GNU General Public License v3.0

 If you encouter any error or have questions,
   send an email to the developer

 You can download soruce code and latest version of\n  program at github:
   https://github.com/mir-am/easydata
   """


class TopWin(Toplevel):

    def __init__(self, master, title, message, width, height, wm_del=False):

        Toplevel.__init__(self, master, width=width, height=height)

        self.title(title)
        self.resizable(0, 0)
        self.grab_set()
        self.transient()

        if wm_del:

            self.protocol("WM_DELETE_WINDOW", self.close_win)

        Message(self, text=message, width=320,
                font=tkFont.Font(family="Arial", size=9, weight=tkFont.BOLD)).pack(side=TOP)

    def close_win(self):

        tkmsg.showwarning('Warning!', 'The Program has not fineshed its job yet!')


# widgets created in this function
def program_gui():

    #  creating Top-Level Window

    top_level_win = Tk()
    top_level_win.title("Easy Data By MiR.am")
    top_level_win.minsize(580, 405)
    top_level_win.maxsize(580, 405)
    top_level_win.resizable(0, 0)

    # creaing menu for top-level window

    top_win_menu = Menu(top_level_win)
    top_level_win.config(menu=top_win_menu)

    main_menu = Menu(top_win_menu, tearoff=0)

    main_menu.add_command(label='About', underline=0, command=lambda: TopWin(top_level_win,
                                                        "About Program", about, 420, 250))
    main_menu.add_command(label='Exit', underline=0, command=sys.exit)

    top_win_menu.add_cascade(label="Main Menu", underline=0, menu=main_menu)

    # creating a label for helping users
    help_label = Label(top_level_win, text="Select Stations with Ctrl or Shift key\n"
                                           " Make sure that you are connected to the internet!",
                       font=tkFont.Font(family="Arial", size=9, weight=tkFont.BOLD))
    help_label.grid(row=0, column=1, pady=5)

    # creating Label Frame for Stations
    st_label = LabelFrame(top_level_win, text="Stations")
    st_label.grid(row=1, column=0, rowspan=2, padx=3, pady=3)

    st_list = Listbox(st_label, width=20, height=20, selectmode=EXTENDED)  # list box for inserting stations

    scroll_bar = Scrollbar(st_label, orient=VERTICAL)  # scroll bar for list box

    scroll_bar.config(command=st_list.yview)
    st_list.config(yscrollcommand=scroll_bar.set)

    st_list.grid(row=1, column=0, padx=4, pady=4)
    scroll_bar.grid(row=1, column=1, sticky=NE + SE)

    # inserting stations names in list box
    for label in range(misc.stations.__len__()):

        st_list.insert(label, misc.stations[label].values()[0] + " - " +misc.stations[label].keys()[0])


    # creating label frame for variables
    var_label = LabelFrame(top_level_win, text="Select Variable")
    var_label.grid(row=1, column=1, padx=3, pady=3, sticky=N)

    cb_var_list = []

    for cb_box in range(misc.var_names.__len__()):

        cb_values = IntVar()

        Checkbutton(var_label, text=misc.var_names[cb_box + 1],
                    variable=cb_values, onvalue=cb_box + 1).grid(row=cb_box, column=0, sticky=W)

        cb_var_list.append(cb_values)

    # creating label frame for data types

    data_label = LabelFrame(top_level_win, text="Data Types")
    data_label.grid(row=1, column=2, padx=3, pady=3, sticky=N)

    cb_dt_list = []

    for c_box in range(misc.data_type_name.__len__()):

        cb_values = IntVar()

        Checkbutton(data_label, text=misc.data_type_name[c_box], variable=cb_values,
                    onvalue=c_box + 1).grid(row=c_box, column=0, sticky=W)

        cb_dt_list.append(cb_values)

    # creating label frame for save location

    save_label = LabelFrame(top_level_win, text="Save Location")
    save_label.grid(row=2, column=1, sticky=NW)

    # a button for choosing a path for saving data
    # a variable for user path
    global save_path
    save_path = None

    def ask_dir():

        global save_path
        save_path = tkFileDialog.askdirectory().replace('/', "\\")
        address_label.config(text= "Address: " + save_path)

    save_button = Button(save_label, text="Save", width=10, height=1, command=ask_dir)
    save_button.grid(row=0, column=0, padx=8, pady=8)

    address_label = Label(save_label)
    address_label.grid(row=1, column=0, padx=5, pady=5)

    # creating label frame for processing data section

    pd_label = LabelFrame(top_level_win, text="Getting Data")
    pd_label.grid(row=2, column=2)

    # a list for saving user data
    global user_data_type
    user_data_type = []

    # a list for saving user variables
    global user_var_list
    user_var_list = []

    # get data for user
    def get_data():

        if not misc.get_data_flag:

            global user_data_type
            global user_var_list
            user_data_type = []
            user_var_list = []

            for value in cb_dt_list:

                if value.get() != 0:

                    user_data_type.append(str(value.get()))

            for value in cb_var_list:

                if value.get() != 0:

                    user_var_list.append(value.get())

            # check that user select options
            if (0 < st_list.curselection().__len__() <= 30) and \
                user_var_list.__len__() != 0 and user_data_type.__len__() != 0\
                    and save_path is not None:

                if tkmsg.askyesno("Validation", "Are you sure that you want to get data with"
                                                " these options?"):

                    pd_thread = Thread(target=getdata.process_user_rq, args=(user_data_type, st_list.curselection(),
                                                                             user_var_list, save_path))

                    pd_thread.start()

                    tkmsg.showinfo("Please Wait...", "The Program is making your data ready!\n The Windows"
                                   " Explorer will open your data when the data is ready")


                else:

                    user_data_type = []
                    user_var_list = []

            else:

                tkmsg.showwarning("Warning!", "You should select all options such as stations,\n"
                                  " variables, data types and save location")

        else:

            tkmsg.showwarning("Warning!", "Please Wait...\n The Program is busy now! "
                                          "Because of your last data reqeust!")

    # creating a button for getting data from get data module
    gd_button = Button(pd_label, text="Start!", width=10, height=1, command=get_data)
    gd_button.grid(row=0, column=0, padx=5, pady=5)

    top_level_win.mainloop()


sourcecode/getdata.py
__author__ = 'MiR.am'

# program name: EasyData
# developer: MiR.am
# date: July 11, 2015
# version: 0.5 Alpha
# License: GNU General Public License v3.0


# in this module we get data from chaharmahalmet website

# we use urllib to download content of webpage
# using parse package for parsing string
# we use beatifulsoup module for parsing the HTML
# os module imported for make directory
# time module imported for measuring running time

from colorama import Fore, init
from bs4 import BeautifulSoup
from datetime import datetime
from time import time as r_time
import urllib
import os
import subprocess
import misc
import excelfile


# download data from web page
def dl_data(urlname):

    try:

        return urllib.urlopen(urlname).read()

    except Exception as error:

        print (Fore.RED + "Error: Make sure that you are connected to the internet!\n"
               + str(error))


# get data from downloaded page
def parse_htmlpage(htmldoc):

    try:

        return BeautifulSoup(htmldoc).find_all('pre')[1].contents[0]

    except Exception as error:

        print (Fore.RED + "Error: Could not get data!\n" + str(error))


# saving data in textfile
def save_textfile(filename, content):

    try:

        open(filename, 'w').writelines(content + "\n")

    except Exception as error:

        print (Fore.RED + "Error: Could not save data in textfile!\n" + str(error))

# reading data from text file
def read_data(filename):

    try:

        return open(filename, 'r').readlines()

    except Exception as error:

        print (Fore.RED + "Error: Could not read data from text file!\n" + str(error))


# getting station name, lat, lon, elevation, years,
# mean and sd from string, actually we get long-term data. e.g 10 years data
def ext_lt_data(data):

    try:

        data_ext = {"stationname": None, "lat": None, "lon": None, "elevation": None,
                    "Mean": None, "SD": None, "years": 0}

        for line in data:

            if "STATION" in line:

                data_ext["stationname"] = line.strip()[line.index("STATION")
                                                          + 5: 25].strip()

            elif "LATITUDE" in line:

                # Converting to decimal degree
                # using round to make numbers only 2 digits
                data_ext["lat"] = round(float(line.split()[1:3][0]) + (float(line.split()[1:3][1])
                                                                       / 60), 2)

            elif "LONGITUDE" in line:

                data_ext["lon"] = round(float(line.split()[1:3][0]) + (float(line.split()[1:3][1])
                                                                       / 60), 2)

            elif "ELEVATION" in line:

                data_ext["elevation"] = float(line.split()[1])

            elif "MEAN" in line:

                # using try/except block for preventing error during run time
                try:

                    data_ext["Mean"] = float(line.split()[-1])

                except ValueError:

                    pass

            elif "SD" in line:

                data_ext["SD"] = float(line.split()[-1])

            # this section count number of aviable years for a station
            if line.rstrip():  # ingnoring blank lines

                # this if checks count number of years with data
                if line.strip().split()[0].isdigit():

                    data_ext["years"] += 1

        return data_ext

    except Exception as error:

        print (Fore.RED + "Could not extract long-term data from text file!\n" + str(error))


# getiing anually data from stations
def ext_data_anually(data):

    try:

        data_ext_anually = {"years": [], "data": []}

        for line in data:

            if line.rstrip():

                # storing a row of data
                row_data = line.strip().split()

                if row_data[0].isdigit():

                    # adding years to the list
                    data_ext_anually["years"].append(int(row_data[0]))

                    try:
                        # adding data to the list
                        data_ext_anually["data"].append(float(row_data[-1]))

                    except ValueError:

                        data_ext_anually["data"].append(row_data[-1])

        return data_ext_anually

    except Exception as error:

        print (Fore.RED + "Could not extract Anually data from text file!\n" + str(error))


# getting monthly data from stations
def ext_data_monthly(data):

    try:

        # for storing monthly data
        data_ext = None

        for line in data:

            if "MEAN" in line.strip()[0:4]:

                data_ext = line.strip().split()[1:-1]

                break

        # converting string to float numbers
        for data in range(data_ext.__len__()):

            try:

                data_ext[data] = float(data_ext[data])

            except ValueError:

                pass

        return data_ext

    except Exception as error:

        print (Fore.RED + "Could not extract Monthly from text file!\n" + str(error))


# extract full data from text files
def ext_full_data(data):

    try:

        data_ext = []

        for line in data:

            if "YEAR" in line:

                for line_new in data[data.index(line):]:

                    if line_new.rstrip():

                        list_temp = line_new.split()

                        for index in range(list_temp.__len__()):

                            try:

                                list_temp[index] = float(list_temp[index])

                            except ValueError:

                                pass

                        data_ext.append(list_temp)

                break

        return data_ext

    except Exception as error:

        print(Fore.RED + "Error: Could not extract full data from text file!\n" + str(error))


# make directories in file system
def make_dir(location):

    try:

        # storing path for saving text files and excel files
        save_path = []

        # make directories for saving files
        for path in [location + "\mydata-" + datetime.today().strftime("%Y-%m-%d-%H-%M"), location +
                    "\mydata-" + datetime.today().strftime("%Y-%m-%d-%H-%M") + "\\textfiles", location +
                    "\mydata-" + datetime.today().strftime("%Y-%m-%d-%H-%M") + "\excelfiles"]:

            os.makedirs(path)

            save_path.append(path)

        return save_path

    except Exception as error:

        print (Fore.RED + "Error: Could not make directories in your computer!\n" + str(error))


# save data as text files for user
def save_data_tf(st_code=[], user_var=[], save_loc=[]):

    try:

        for code in st_code:

            for var in user_var:

                # URL structure: http://www.chaharmahalmet.ir/stat/archive/
                # iran/provincename/stationname/variablename.asp

                path = save_loc[1] + "\\" + misc.stations[int(code)].keys()[0].capitalize() + "-" +\
                       misc.short_var_names[var] + ".txt"

                save_textfile(path,
                parse_htmlpage(dl_data("http://www.chaharmahalmet.ir/stat/archive/iran/%s/%s/%s.asp"
                % (misc.stations[int(code)].values()[0], misc.stations[int(code)].keys()[0], misc.variables[var]))))

        print (Fore.GREEN + "Data saved as text files!\n")

    except Exception as error:

        print (Fore.RED + "Error: Could not save data as text files!\n" + str(error))


# make long-term data ready for users
def ps_lt_data(st_code=[], user_var=[], save_loc=None):

    try:

        # stations long-term data will store in this variable
        stations_data = {"st_names": [], "lat": [], "lon": [], "elevation": [], "years": [], "tavg": [],
                         "tmin": [], "tmax": [], "rh": [], "pre": [], "dust": [], "wind": [], "sh": []}

        for code in st_code:

            # using this variable for previnting duplicate values for lists
            new_st = True

            for var in user_var:

                path = save_loc[1] + "\\" + misc.stations[int(code)].keys()[0].capitalize() + "-" +\
                       misc.short_var_names[var] + ".txt"

                store_data = ext_lt_data(read_data(path))

                if new_st:

                    stations_data["st_names"].append(store_data["stationname"])

                    stations_data["lat"].append(store_data["lat"])

                    stations_data["lon"].append(store_data["lon"])

                    stations_data["elevation"].append(store_data["elevation"])

                    stations_data["years"].append(store_data["years"])

                    new_st = False

                stations_data[misc.short_var_names[var]].append(store_data["Mean"])

        excelfile.ct_lt_excel(stations_data, save_loc[2])

        print (Fore.GREEN + "Long-term Excel data is now ready!\n")

    except Exception as error:

        print (Fore.RED + "Error: Could not Creating long-term excel files!\n " + str(error))


# make anually data ready for users
def ps_data_anually(st_code=[], user_var=[], save_loc=None):

    try:

        # stations anually
        # example: st_data_anually = {"NAEIN": {"year": [2003, 2004], "pre": [23.4, 120] },
        #  "YAZD": {"years": [2001, 2002], "tmax": [28, 32] } }
        st_data_anually = {}

        for code in st_code:

            st_data_anually[misc.stations[int(code)].keys()[0]] = {}

            new_st = True

            for var in user_var:

                path = save_loc[1] + "\\" + misc.stations[int(code)].keys()[0].capitalize() + "-" +\
                       misc.short_var_names[var] + ".txt"

                store_data = ext_data_anually(read_data(path))

                if new_st:

                    st_data_anually[misc.stations[int(code)].keys()[0]]["years"] = store_data["years"]

                    new_st = False

                st_data_anually[misc.stations[int(code)].keys()[0]][misc.short_var_names[var]] = store_data["data"]

        excelfile.ct_anually_excel(st_data_anually, save_loc[2])

        print (Fore.GREEN + "Anually data as excel files is now ready!\n")

    except Exception as error:

        print (Fore.RED + "Error: Could not creating anually data!\n" + str(error))


# make monthly data ready for users
def ps_data_monthly(st_code=[], user_var=[], save_loc=None):

    try:

        # monthly data
        # exmaple: st_data_monthly = {"ARAK": {"tmin": [10, 12]}}
        st_data_monthly = {}

        for code in st_code:

            st_data_monthly[misc.stations[int(code)].keys()[0]] = {}

            for var in user_var:

                path = save_loc[1] + "\\" + misc.stations[int(code)].keys()[0].capitalize() + "-" +\
                       misc.short_var_names[var] + ".txt"

                store_data = ext_data_monthly(read_data(path))

                st_data_monthly[misc.stations[int(code)].keys()[0]][misc.short_var_names[var]] = store_data

        excelfile.ct_monthly_excel(st_data_monthly, save_loc[2])

        print (Fore.GREEN + "Monthly data as excel files is now ready!\n")

    except Exception as error:

        print (Fore.RED + "Error: Could not creating Monthly data!\n" + str(error))


# make full data ready for users
def ps_full_data(st_code=[], user_var=[], save_loc=None):

    try:

        # storing full data
        # example: {"tabriz": {"pre": [[...], [...]] } }
        st_full_data = {}

        for code in st_code:

            st_full_data[misc.stations[int(code)].keys()[0]] = {}

            for var in user_var:

                path = save_loc[1] + "\\" + misc.stations[int(code)].keys()[0].capitalize() + "-" +\
                       misc.short_var_names[var] + ".txt"

                store_data = ext_full_data(read_data(path))

                st_full_data[misc.stations[int(code)].keys()[0]][misc.short_var_names[var]] = store_data

        excelfile.ct_fulldata_excel(st_full_data, save_loc[2])

        print(Fore.GREEN + "Full data as excel files is now ready!\n")

    except Exception as error:

        print("Error: Could not creating full data from text files\n" + str(error))

# prcoess user request
def process_user_rq(user_choices=[], st_code=[], user_var=[], save_loc=None):

    misc.get_data_flag = True

    time = r_time()

    path = make_dir(save_loc)

    save_data_tf(st_code, user_var, path)

    # "1" is long-term data
    if "1" in user_choices:

        ps_lt_data(st_code, user_var, path)

    # "2" is Anually data
    if "2" in user_choices:

        ps_data_anually(st_code, user_var, path)

    # "3" is Monthly data
    if "3" in user_choices:

        ps_data_monthly(st_code, user_var, path)

    # "4" is full data
    if "4" in user_choices:

        ps_full_data(st_code, user_var, path)

    print (Fore.GREEN + "You can see your data at this address: %s\nyour request finished in %.2f"
            " seconds\n" % (path[0], r_time() - time))

    subprocess.call('explorer ' + "\"" + path[0] + "\"", shell=True)

    misc.get_data_flag = False

sourcecode/excelfile.py
__author__ = 'MiR.am'

# program name: EasyData
# developer: MiR.am
# date: June 14, 2015
# version: 0.4 Alpha
# License: GNU General Public License v3.0

# data will convert to excel file with help of this module.

# using xlsx writer to convert data to excel file
from datetime import datetime
from colorama import Fore
import xlsxwriter


# header names for excel file
headers_name = ("st_names", "lat", "lon", "elevation", "years", "pre", "tavg", "tmin", "tmax",
                "rh", "dust", "wind", "sh")


# month names for monthly data
month_names = ("Months", "JAN.", "FEB.", "MAR.", "APR.", "MAY", "JUNE", "JULY", "AUG.", "SEP.",
                  "OCT.", "NOV.", "DEC.")


# this method convert data to long-terem excel data
def ct_lt_excel(st_data={}, path=None):

    try:

        # creating an excel file
        workbook = xlsxwriter.Workbook(path + "\long-term-data" + datetime.today()
                                       .strftime("%Y-%m-%d-%H-%M") + ".xlsx")
        # creating a sheet
        worksheet = workbook.add_worksheet()

        global row
        column = 0

        for header in headers_name:

            # only variables with value should be written in excel file
            if st_data[header].__len__() > 0:

                # row should be zero when writing to the next column
                row = 0

                worksheet.write(row, column, header)

                row += 1

                for data in st_data[header]:

                    worksheet.write(row, column, data)

                    row += 1

                column += 1

        # file should be closed
        workbook.close()

    except Exception as error:

        print (Fore.RED + "Error: Could not convert long-term data to excel file!\n" + str(error))


# this method convert data to anually excel data
def ct_anually_excel(st_data={}, path=None):

    try:

        # creating an excel file
        workbook = xlsxwriter.Workbook(path + "\\anually-data" + datetime.today()
                                           .strftime("%Y-%m-%d-%H-%M") + ".xlsx")

        global row
        column = 0

        for st_name in st_data.keys():

            worksheet = workbook.add_worksheet(st_name.capitalize())

            for var in headers_name:

                if var in st_data[st_name].keys():

                    row = 0

                    worksheet.write(row, column, var)

                    row += 1

                    for data in st_data[st_name][var]:

                        worksheet.write(row, column, data)

                        row += 1

                    column += 1

            row = 0
            column = 0

        workbook.close()

    except Exception as error:

        print(Fore.RED + "Error: Could not convert anually data to excel file!\n" + str(error))


# this method convert data to monthly excel data
def ct_monthly_excel(st_data={}, path=None):

    try:

        # creating an excel file
        workbook = xlsxwriter.Workbook(path + "\\monthly-data" + datetime.today()
                                           .strftime("%Y-%m-%d-%H-%M") + ".xlsx")

        global row
        column = 0
        row = 0

        for st_name in st_data.keys():

            worksheet = workbook.add_worksheet(st_name.capitalize())

            for header in month_names:

                worksheet.write(row, 0, header)

                row += 1

            row = 0
            column = 1

            for var in st_data[st_name].keys():

                worksheet.write(row, column, var)

                row += 1

                for data in st_data[st_name][var]:

                    worksheet.write(row, column, data)

                    row += 1

                column += 1
                row = 0

        workbook.close()

    except Exception as error:

        print(Fore.RED + "Error: Could not convert monthly data to excel file!\n" + str(error))


# this method convert data to full data excel files
def ct_fulldata_excel(st_data={}, path=None):

    try:

        row = 0
        column = 0

        for st_name in st_data.keys():

            workbook = xlsxwriter.Workbook(path + "\\fulldata-" + st_name.capitalize() +
                                           datetime.today().strftime("%H-%M") + ".xlsx")

            for var in st_data[st_name].keys():

                worksheet = workbook.add_worksheet(var)

                for list_2d in st_data[st_name][var]:

                    for data in list_2d:

                        worksheet.write(row, column, data)

                        column += 1

                    row += 1
                    column = 0

                column = 0
                row = 0

    except Exception as error:

        print(Fore.RED + "Error: Could not convert full data to excel file!\n" + str(error))


sourcecode/misc.py
__author__ = 'MiR.am'

# program name: EasyData
# developer: MiR.am
# date: June 14, 2015
# version: 0.4 Alpha
# License: GNU General Public License v3.0

# in this module, variables and stations has defined


var_names = {1: "AVERAGE OF MEAN DAILY TEMPERATURE IN C",
             2: "AVERAGE OF MINIMUM TEMPERATURE IN C",
             3: "AVERAGE OF MAXIMUM TEMPERATURE IN C",
             4: "AVERAGE OF RELATIVE HUMIDITY IN PERCENT",
             5: "MONTHLY TOTAL OF PRECIPITATION IN MM",
             6: "NO. OF DAYS WITH DUST",
             7: "AVERAGE OF WIND SPEED IN KNOTS",
             8: "MONTHLY TOTAL OF SUNSHINE HOURS"}


# short names for variables
short_var_names = {1: "tavg", 2: "tmin", 3: "tmax", 4: "rh", 5: "pre", 6: "dust",
                   7: "wind", 8: "sh"}


# every variabe has a code
variables = {1: "5", 2: "2", 3: "3", 4: "14", 5: "25",
             6: "33", 7: "36", 8: "42"}



# statations and provinces stored in dictionary
# Beautiful Soup 4 module used for getting station names from chaharmahal
# web site. it saves time and life!!
# given a code to a station. between 0 to 202. e.g SHAHREKO's code(index) is 0

stations = ({u'SHAHREKO': "chb"}, {u'BOROOJEN': "chb"}, {u'KOOHRANG': "chb"}, {u'LORDEGAN': "chb"},
            {u'TABRIZ': "azs"}, {u'AHAR': "azs"}, {u'BONAB': "azs"}, {u'JOLFA': "azs"}, {u'SARAB': "azs"},
            {u'SAHAND': "azs"}, {u'KALEIBAR': "azs"}, {u'MARAGHEH': "azs"}, {u'MARAND': "azs"}, {u'MIANEH': "azs"},
            {u'OROOMIEH': "azg"}, {u'KHOY': "azg"}, {u'MAHABAD': "azg"}, {u'MAKOO': "azg"}, {u'PIRANSHR': "azg"},
            {u'SALMAS': "azg"}, {u'SARDASHT': "azg"}, {u'TAKAB': "azg"}, {u'ARDEBIL': "ard"}, {u'KHALKHAL': "ard"},
            {u'MESHKINS': "ard"}, {u'PARSABAD': "ard"}, {u'ESFAHAN': "esf"}, {u'ARDESTAN': "esf"}, {u'DARAN': "esf"},
            {u'KABOOTAR': "esf"}, {u'KASHAN': "esf"}, {u'KHORBIAB': "esf"}, {u'MEIMEH': "esf"}, {u'GOLPAIGN': "esf"},
            {u'NAEIN': "esf"}, {u'NATANZ': "esf"}, {u'SHAHREZA': "esf"}, {u'SHARGHES': "esf"}, {u'KARAJ': "alb"},
            {u'ILAM': "ila"}, {u'DEHLORAN': "ila"}, {u'EIVANGHA': "ila"}, {u'MEHRAN': "ila"}, {u'DAREHSHA': "ila"},
            {u'BUSHEHR': "bos"}, {u'BUSHCOST': "bos"}, {u'BANDARDA': "bos"}, {u'BANDARDL': "bos"}, {u'KANGANJA': "bos"},
            {u'TEHRAN': "teh"}, {u'SHOMALTE': "teh"}, {u'CHITGAR': "teh"}, {u'ABALI': "teh"}, {u'JEOPHYSI': "teh"},
            {u'FIROUZKO': "teh"}, {u'FIROUZPL': "teh"}, {u'DOUSHANT': "teh"}, {u'EMAMFORO': "teh"}, {u'BIRJAND': "khj"},
            {u'GHAEN': "khj"}, {u'KHOREBIR': "khj"}, {u'MASHHAD': "khr"}, {u'BOSHROOY': "khr"}, {u'FERDOUS': "khr"},
            {u'GHOOCHAN': "khr"}, {u'GOLMAKAN': "khr"}, {u'GONABAD': "khr"}, {u'KASHMAR': "khr"}, {u'NEYSHABO': "khr"},
            {u'SABZEVAR': "khr"}, {u'SARAKHS': "khr"}, {u'TORBATEH': "khr"}, {u'TORBATEJ': "khr"}, {u'FARIMAN': "khr"},
            {u'BOJNURD': "khs"}, {u'AHWAZ': "khz"}, {u'ABADAN': "khz"}, {u'AGHAJARI': "khz"}, {u'BANDARMA': "khz"},
            {u'BEHBAHAN': "khz"}, {u'BOSTAN': "khz"}, {u'DEZFUL': "khz"}, {u'HENDIJAN': "khz"}, {u'IZEH': "khz"},
            {u'MASJEDSO': "khz"}, {u'OMIDIYEH': "khz"}, {u'RAMHORMO': "khz"}, {u'SAFIABAD': "khz"}, {u'SHOSHTAR': "khz"},
            {u'ZANJAN': "zan"}, {u'MAHNESHA': "zan"}, {u'KHODABAN': "zan"}, {u'KHORMDAR': "zan"}, {u'SEMNAN': "sem"},
            {u'BIARJAMA': "sem"}, {u'SHAHROUD': "sem"}, {u'GARMSAR': "sem"}, {u'DAMGHAN': "sem"}, {u'ZAHEDAN': "sis"},
            {u'CHABAHAR': "sis"}, {u'IRANSHAR': "sis"}, {u'KENARAK': "sis"}, {u'KHASH': "sis"}, {u'SARAVAN': "sis"},
            {u'ZABOL': "sis"}, {u'ZAHAK': "sis"}, {u'NIKSHAHR': "sis"}, {u'KHAVAF': "sis"}, {u'SHIRAZ': "far"},
            {u'ABADEH': "far"}, {u'DARAB': "far"}, {u'FASSA': "far"}, {u'LAMERD': "far"}, {u'LAR': "far"},
            {u'NEIRIZE': "far"}, {u'SADDOROU': "far"}, {u'ZARGHAN': "far"}, {u'EGHLIDEF': "far"}, {u'SAFASHAR': "far"},
            {u'KAZEROON': "far"}, {u'GHOM': "qom"}, {u'SALAFCHE': "qom"}, {u'GHAZVIN': "ghz"}, {u'AVAJ': "ghz"},
            {u'SANANDAJ': "kor"}, {u'BANEH': "kor"}, {u'BIJAR': "kor"}, {u'GHORVEH': "kor"}, {u'MARIVAN': "kor"},
            {u'SAGHEZ': "kor"}, {u'ZARINEHO': "kor"}, {u'KERMAN': "ken"}, {u'BAFT': "ken"}, {u'BAM': "ken"},
            {u'KAHNOUJ': "ken"}, {u'MIANDEHJ': "ken"}, {u'RAFSANJN': "ken"}, {u'SHAHRBAB': "ken"}, {u'SIRJAN': "ken"},
            {u'SHAHDAD': "ken"}, {u'LALEHZAR': "ken"}, {u'KERMANSH': "keh"}, {u'ESLAMABA': "keh"}, {u'KANGAVAR': "keh"},
            {u'RAVANSAR': "keh"}, {u'SARARUDK': "keh"}, {u'SARPOLZO': "keh"}, {u'YASOUJ': "koh"}, {u'DOGONBAD': "koh"},
            {u'GORGAN': "gol"}, {u'GONBADEG': "gol"}, {u'MORAVEHT': "gol"}, {u'RASHT': "gil"}, {u'ANZALI': "gil"},
            {u'ASTARA': "gil"}, {u'MANJIL': "gil"}, {u'JIRANDEH': "gil"}, {u'KHORAMAB': "lor"}, {u'BROUJERD': "lor"},
            {u'ALGODARZ': "lor"}, {u'ALESHTAR': "lor"}, {u'AZNA': "lor"}, {u'DOROUD': "lor"}, {u'KOUDASHT': "lor"},
            {u'NORABADE': "lor"}, {u'POLDOKHT': "lor"}, {u'SARI': "maz"}, {u'RAMSAR': "maz"}, {u'BABOLSAR': "maz"},
            {u'NOUSHAHR': "maz"}, {u'GHAEMSHR': "maz"}, {u'AMOL': "maz"}, {u'SIAHBISH': "maz"}, {u'KIASAR': "maz"},
            {u'ARAK': "mar"}, {u'SAVEH': "mar"}, {u'TAFRESH': "mar"}, {u'KHOMEIN': "mar"}, {u'MAHALLAT': "mar"},
            {u'BANDARAB': "hor"}, {u'BANDARLN': "hor"}, {u'MINAB': "hor"}, {u'JASK': "hor"}, {u'HAJIABAD': "hor"},
            {u'ABOMOOSA': "hor"}, {u'GHESHM': "hor"}, {u'KISH': "hor"}, {u'SIRI': "hor"}, {u'LAVAN': "hor"},
            {u'HAMEDANF': "ham"}, {u'HAMEDANO': "ham"}, {u'MALAYER': "ham"}, {u'NAHAVAND': "ham"}, {u'TOYSERKA': "ham"},
            {u'YAZD': "yaz"}, {u'ANAR': "yaz"}, {u'BAFGH': "yaz"}, {u'MARVAST': "yaz"}, {u'ROBATPO': "yaz"},
            {u'TABASS': "yaz"}, {u'AGHDA': "yaz"}, {u'GARIZ': "yaz"})



data_type_name = ("Long-Term Data", "Anually Data", "Monthly Data", "Full Data")

# a flag for critical section of get data module.
# because user can request data at one time.

get_data_flag = False


# print full names of variables for users.
# because they shoud ensure
def print_var(user_var=[]):

    for var in user_var:

        print var + " -> " + var_names[var]


# print names of stations for users
# because they shoud ensure
def print_st_names(st_codes=[]):

    for code in st_codes:

        # code should converted to the integer for list index
        print "%s -> " % code + stations[int(code)].keys()[0].capitalize()


# for index in range(stations.__len__()):
#
#     print str(index) + " -> " , stations[index]
